File: engine/src/theseus_engine/guidance.py
"""Guidance ("Ariadne mode"): turn a planned path into egocentric cues.

The hero use case is a guided human, so the engine's real output is not a
polyline — it is a stream of instructions a person can act on without
seeing the map: "straight 3 m", "turn left 45°", "you're off the thread".
On-device these cues drive haptics (CoreHaptics), spatial audio (a PHASE
beacon placed at the lookahead point) and speech; here they are a typed
value that tests can assert on, which is exactly why the geometry
(especially turn SIGN: +angle = turn left, see geometry.py conventions)
is pinned down on Windows before any iPhone is involved.

Mechanics: pure-pursuit style. Project the pose onto the path, hold a
lookahead point a fixed arclength ahead, and report the bearing error to
it. Cross-track distance beyond `corridor_max` means the human has left
the corridor -> OFF_ROUTE, which the controller turns into a replan from
wherever they actually are.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .geometry import Vec, bearing, dist, point_along, project_point_segment, wrap_angle
from .grid import OccupancyGrid

STRAIGHT, TURN_LEFT, TURN_RIGHT, ARRIVE, OFF_ROUTE = (
    "straight", "turn_left", "turn_right", "arrive", "off_route")


@dataclass
class GuidanceCue:
    kind: str
    distance: float        # m: to next turn (straight), to path (off_route), to goal (arrive)
    angle_deg: float       # bearing error to lookahead point; + = turn left
    cross_track: float     # m off the path centerline
    corridor: float        # local corridor width (2 x clearance), m
    target: Vec            # lookahead point in world coords (steering aims here)


class GuidanceFollower:
    def __init__(self, path: list[Vec], lookahead: float = 0.9,
                 corridor_max: float = 0.9, arrive_radius: float = 0.45,
                 turn_thresh_deg: float = 22.0, vertex_turn_deg: float = 28.0):
        if len(path) < 1:
            raise ValueError("empty path")
        self.path = path
        self.lookahead = lookahead
        self.corridor_max = corridor_max
        self.arrive_radius = arrive_radius
        self.turn_thresh_deg = turn_thresh_deg
        self.vertex_turn_deg = vertex_turn_deg

    def _project(self, p: Vec) -> tuple[Vec, int, float, float]:
        """Nearest point on the path: (point, segment index, t, distance)."""
        if len(self.path) == 1:
            q = self.path[0]
            return q, 0, 0.0, dist(p, q)
        best = None
        for i in range(len(self.path) - 1):
            q, t = project_point_segment(p, self.path[i], self.path[i + 1])
            d = dist(p, q)
            if best is None or d < best[3]:
                best = (q, i, t, d)
        return best

    def _dist_to_next_turn(self, seg_i: int, t: float) -> float:
        """Arclength from the projection to the first vertex where the path
        bends by more than vertex_turn_deg (else to the goal)."""
        proj = point_along(self.path, seg_i, t, 0.0)
        acc = dist(proj, self.path[seg_i + 1]) if seg_i + 1 < len(self.path) else 0.0
        for j in range(seg_i + 1, len(self.path) - 1):
            d_in = bearing(self.path[j - 1] if j - 1 >= 0 else proj, self.path[j])
            d_out = bearing(self.path[j], self.path[j + 1])
            if abs(math.degrees(wrap_angle(d_out - d_in))) > self.vertex_turn_deg:
                return acc
            acc += dist(self.path[j], self.path[j + 1])
        return acc

    def cue(self, grid: OccupancyGrid, pose: tuple[float, float, float]) -> GuidanceCue:
        x, y, th = pose
        p = (x, y)
        goal = self.path[-1]
        if dist(p, goal) <= self.arrive_radius:
            return GuidanceCue(ARRIVE, dist(p, goal), 0.0, 0.0,
                               2.0 * grid.clearance(grid.world_to_cell(goal)), goal)
        proj, seg_i, t, cross = self._project(p)
        corridor = 2.0 * grid.clearance(grid.world_to_cell(proj))
        if cross > self.corridor_max:
            err = math.degrees(wrap_angle(bearing(p, proj) - th))
            return GuidanceCue(OFF_ROUTE, cross, err, cross, corridor, proj)
        look = point_along(self.path, seg_i, t, self.lookahead) if len(self.path) > 1 else goal
        err = math.degrees(wrap_angle(bearing(p, look) - th))
        if abs(err) < self.turn_thresh_deg:
            return GuidanceCue(STRAIGHT, self._dist_to_next_turn(seg_i, t),
                               err, cross, corridor, look)
        kind = TURN_LEFT if err > 0 else TURN_RIGHT
        return GuidanceCue(kind, dist(p, look), err, cross, corridor, look)

File: apps/navcore/Sources/NavCore/Guidance.swift
// Guidance ("Ariadne mode") — port of engine guidance.py: turn a
// planned path into egocentric cues a person can act on without seeing
// the map. On-device these drive haptics, audio and speech; here they
// are a typed value tests can assert on — especially the turn SIGN
// (+angle = turn LEFT, the project-wide convention).
//
// Mechanics: pure pursuit. Project the pose onto the path, hold a
// lookahead point a fixed arclength ahead, report the bearing error.
// Cross-track beyond corridorMax means the human left the corridor.

public enum CueKind: String, Sendable {
    case straight
    case turnLeft = "turn_left"
    case turnRight = "turn_right"
    case arrive
    case offRoute = "off_route"
}

public struct GuidanceCue: Sendable {
    public var kind: CueKind
    public var distance: Double     // m: to next turn / to path / to goal
    public var angleDeg: Double     // bearing error to lookahead; + = LEFT
    public var crossTrack: Double   // m off the path centerline
    public var corridor: Double     // local corridor width (2x clearance), m
    public var target: Vec          // lookahead point (steering aims here)
}

public final class GuidanceFollower {
    public let path: [Vec]
    public let lookahead: Double
    public let corridorMax: Double
    public let arriveRadius: Double
    public let turnThreshDeg: Double
    public let vertexTurnDeg: Double

    public init(path: [Vec], lookahead: Double = 0.9,
                corridorMax: Double = 0.9, arriveRadius: Double = 0.45,
                turnThreshDeg: Double = 22.0, vertexTurnDeg: Double = 28.0) {
        precondition(!path.isEmpty, "empty path")
        self.path = path
        self.lookahead = lookahead
        self.corridorMax = corridorMax
        self.arriveRadius = arriveRadius
        self.turnThreshDeg = turnThreshDeg
        self.vertexTurnDeg = vertexTurnDeg
    }

    /// Nearest point on the path: (point, segment index, t, distance).
    func project(_ p: Vec) -> (Vec, Int, Double, Double) {
        if path.count == 1 {
            let q = path[0]
            return (q, 0, 0.0, dist(p, q))
        }
        var best: (Vec, Int, Double, Double)? = nil
        for i in 0..<(path.count - 1) {
            let (q, t) = projectPointSegment(p, path[i], path[i + 1])
            let d = dist(p, q)
            if best == nil || d < best!.3 {
                best = (q, i, t, d)
            }
        }
        return best!
    }

    /// Arclength from the projection to the first vertex where the path
    /// bends by more than vertexTurnDeg (else to the goal).
    func distToNextTurn(segI: Int, t: Double) -> Double {
        let proj = pointAlong(path, startI: segI, startT: t, ahead: 0.0)
        var acc = segI + 1 < path.count ? dist(proj, path[segI + 1]) : 0.0
        for j in (segI + 1)..<Swift.max(segI + 1, path.count - 1) {
            let dIn = bearing(from: j - 1 >= 0 ? path[j - 1] : proj,
                              to: path[j])
            let dOut = bearing(from: path[j], to: path[j + 1])
            if abs(wrapAngle(dOut - dIn) * (180.0 / Double.pi))
                > vertexTurnDeg {
                return acc
            }
            acc += dist(path[j], path[j + 1])
        }
        return acc
    }

    public func cue(grid: OccupancyGrid,
                    pose: (Double, Double, Double)) -> GuidanceCue {
        let (x, y, th) = pose
        let p = Vec(x, y)
        let goal = path[path.count - 1]
        if dist(p, goal) <= arriveRadius {
            return GuidanceCue(
                kind: .arrive, distance: dist(p, goal), angleDeg: 0.0,
                crossTrack: 0.0,
                corridor: 2.0 * grid.clearance(grid.worldToCell(goal)),
                target: goal)
        }
        let (proj, segI, t, cross) = project(p)
        let corridor = 2.0 * grid.clearance(grid.worldToCell(proj))
        if cross > corridorMax {
            let err = wrapAngle(bearing(from: p, to: proj) - th)
                * (180.0 / Double.pi)
            return GuidanceCue(kind: .offRoute, distance: cross,
                               angleDeg: err, crossTrack: cross,
                               corridor: corridor, target: proj)
        }
        let look = path.count > 1
            ? pointAlong(path, startI: segI, startT: t, ahead: lookahead)
            : goal
        let err = wrapAngle(bearing(from: p, to: look) - th)
            * (180.0 / Double.pi)
        if abs(err) < turnThreshDeg {
            return GuidanceCue(kind: .straight,
                               distance: distToNextTurn(segI: segI, t: t),
                               angleDeg: err, crossTrack: cross,
                               corridor: corridor, target: look)
        }
        return GuidanceCue(kind: err > 0 ? .turnLeft : .turnRight,
                           distance: dist(p, look), angleDeg: err,
                           crossTrack: cross, corridor: corridor,
                           target: look)
    }
}
